sdk/database/newdbplugin/database.go

```diff
@@ -0,0 +1,155 @@
+package newdbplugin
+
+import (
+	"context"
+	"time"
+)
+
+// Database to manipulate users within an external system (typically a database).
+type Database interface {
+	// Initialize the database plugin. This is the equivalent of a constructor for the
+	Initialize(ctx context.Context, req InitializeRequest) (InitializeResponse, error)
+
+	// NewUser creates a new user within the database. This user is temporary in that it
+	// will exist until the TTL expires.
+	NewUser(ctx context.Context, req NewUserRequest) (NewUserResponse, error)
+
+	// UpdateUser updates an existing user within the database.
+	UpdateUser(ctx context.Context, req UpdateUserRequest) (UpdateUserResponse, error)
+
+	// DeleteUser from the database. This should not error if the user didn't
+	// exist prior to this call.
+	DeleteUser(ctx context.Context, req DeleteUserRequest) (DeleteUserResponse, error)
+
+	// Type returns the Name for the particular database backend implementation.
+	// This type name is usually set as a constant within the database backend
+	// implementation, e.g. "mysql" for the MySQL database backend. This is used
+	// for things like metrics and logging. No behavior is switched on this.
+	Type() (string, error)
+
+	// Close attempts to close the underlying database connection that was
+	// established by the backend.
+	Close() error
+}
+
+// ///////////////////////////////////////////////////////////////////////////
+// Database Request & Response Objects
+// These request and response objects are *not* protobuf types because gRPC does not
+// support all types that we need in a nice way. For instance, gRPC does not support
+// map[string]interface{}. It does have an `Any` type, but converting it to a map
+// requires extensive use of reflection and knowing what types to support ahead of
+// time. Instead these types are made as user-friendly as possible so the conversion
+// between protobuf types and request/response objects is handled by Vault developers
+// rather than needing to be handled by external plugin developers.
+// ///////////////////////////////////////////////////////////////////////////
+
+// ///////////////////////////////////////////////////////
+// Initialize()
+// ///////////////////////////////////////////////////////
+
+// InitializeRequest contains all information needed to initialize a database plugin.
+type InitializeRequest struct {
+	// Config to initialize the database with. This can include things like connection details,
+	// a "root" username & password, etc. This will not include all configuration items specified
+	// when configuring the database. Some values will be stripped out by the database engine
+	// prior to being passed to the plugin.
+	Config map[string]interface{}
```

Comment:
Unfortunately `google.protobuf.Any` isn't really meant as a generic replacement for `interface{}` like we want to use here. Its main use case is to be any protobuf type, not primitive types: [google.protobuf.Any docs](https://pkg.go.dev/google.golang.org/protobuf/types/known/anypb?tab=doc).

If we use the protobuf types in the interface directly this results in a `map[string]Any` config in the request/response objects. This `Any` type will require plugin developers to handle the weirdness of type checking & converting types in their own code rather than having it done in a translation layer automatically. This would result in a lot of copy-pasted code across plugins as well as a poorer user experience especially for non-go developers.

I agree that having a translation layer isn't ideal, but I'd rather have us maintain that translation than put extra onus on plugin developers to understand the `Any` type. Plus I suspect that we won't be updating this interface *that* frequently, though that's just my opinion.